Added proposal for removing Nat in favor of promoted Natural (#2)
Haskell-mouse/core-libraries-proposals: proposals/0001-unify-nat-and-natural.md

---
author: Rinat Stryungis
date-accepted: ""
ticket-url: "https://gitlab.haskell.org/ghc/ghc/-/issues/10776"
implemented: "Yes"
---

This proposal is [discussed at this pull request](https://github.com/haskell-core/core-libraries-proposals/pull/0>).
**After creating the pull request, edit this file again, update the number in
the link, and delete this bold sentence.**

> # Proposal title

Reviewer: ```suggestion
# Unify Nat and Natural
```

> The changes in the proposal remove the separate built-in kind `Nat` 

Reviewer: ```suggestion
Remove the separate built-in kind `Nat` 
```

> in favor of promoted type `Natural`.

Reviewer: ```suggestion
in favor of promoting `Natural`.
```

## Motivation

Existence of separated kind `Nat` and impossibility of promotion 

> of usual `Natural` data type (and data constructors with `Natual` fields)

Reviewer: `Natual` -> `Natural`

looks like an unnecessary and inconvenient complication. 

> It forces to create redundant types with `Nat` fields in their constructors (it makes these types uninhabited in terms)

Reviewer: ```suggestion
It forces us to create redundant types with `Nat` fields in their constructors (it makes these types uninhabited in terms)
```

> only for promotional purposes. This is more cumbersome and less safe than the promotion 

Reviewer: ```suggestion
only for the sake of promotion. This is more cumbersome and less safe than the promotion 
```

Reviewer: By the way, how is it less safe? I think it's just more cumbersome.

Author: It is less safe because now one has to manually create 2 synchronized types and you could make a mistake or forget to change one of them in case of refactoring for example.

Reviewer: There's a technique to avoid this, you would typically write something like this:

```
data MyType' n = MkT X Y n Z

type MyType = MyType' Natural  -- for use in terms
type MyTypeP = MyType' Nat     -- for use in types
```

So you're not really repeating the `X Y Z` part of the declaration.

Author: Yea, but it is inconvenient.... ok you are right :-)

> of the ordinary data type. 

Reviewer: ```suggestion
of ordinary data types.
```

## Proposed Change Specification

> 1. This patch removes the separate kind `Nat` and repaces it by promoted type `Natural`. 

Reviewer: ```suggestion
1. Remove the separate kind `Nat` and replace it with `Natural`. Type-level numeric literals now have the kind `Natural`, i.e. `p :: Proxy (1 :: Nat)` becomes `p :: Proxy (1 :: Natural)`.
```

Reviewer: The patch should be mentioned in the implementation section, but not here. This section should just describe the change itself, so let's avoid phrases such as "This patch removes ..." here.

Also, let's make it clear that the typing rules for type-level literals have changed, i.e. `1 :: Nat` becomes `1 :: Natural`.

> 2. Allows promoting of the data constructors with Natural numbers in their fields. 

Reviewer: I think this follows from (1), so it should go to the "Effect and Interactions" section.

> 3. Module `GHC.TypeNats` now contains type synonym `type Nat = Natural` for the backward compatibility 

Reviewer: ```suggestion
3. Module `GHC.TypeNats` now contains a type synonym `type Nat = Natural` for backwards compatibility.
```

## Examples

> The pretty simple example: 

Reviewer: This line conveys no information, I think it's better to just drop it

Before this change, we had to create a separate type for the promotion

    data C = MkC Natural
    data CP = MkCP Nat

>     type CC = MkCP 1

Reviewer: Let's make it clear why `CP` is needed here by showing which terms don't typecheck.

Before the proposal:

```
data C = MkC Natural
c1 = MkC 1        -- works
type C1 = MkC 1   -- doesn't work

data P = MkP Nat
p1 = MkP 1        -- doesn't work
type P1 = MkP 1   -- works
```

After the proposal:

```
data C = MkC Natural
c1 = MkC 1       -- works
type C1 = MkC 1  -- works
```

But `CP` is uninhabited in terms.

Now the following code will be successfully type-checked:
    
    data C = MkC Natural
    type CC = MkC 1

## Effect and Interactions

1. Type checker plugins which work with the natural numbers now
should use ``naturalTy`` kind instead of removed ``typeNatKind``

2. The separate kind ``Nat`` is removed and now it is just a type synonym for
``Natural``. As a consequence, one must enable ``TypeSynonymInstances``
in order to define instances for ``Nat``.

> ## Costs and Drawbacks

Reviewer: The drawback is that this is a breaking change. Code such as this will break:

```
class C a
instance C Nat
instance C Natural
```

Now these instances are conflicting.

> ## Alternatives

Reviewer: You can state the "null"-alternative here, so to speak, i.e. keeping things as they are:

* Continue to maintain a separate `Nat` kind indefinitely. However, this makes type-level programming with numeric literals less convenient than it could be.

> ## Unresolved Questions

Reviewer: "None at the moment".
